cobra/flux_analysis/deletion.py

```diff
@@ -0,0 +1,348 @@
+        WORKERS = dict(gene=_gene_deletion_worker, reaction=_reaction_deletion_worker)
+        worker_function = WORKERS[entity]
+
+        try:
+            (legacy, solver) = sutil.choose_solver(model, solver, qp=(method == "moma"))
+        except sutil.SolverNotFound:
+            if method == "moma":
+                warn("Cannot use MOMA since no QP-capable solver was found. Falling"
+                     " back to FBA.")
+                (legacy, solver) = sutil.choose_solver(model, solver)
+            else:
+                (err_type, err_val, err_tb) = sys.exc_info()
+                raise_(err_type, err_val, err_tb)  # reraise for Python2&3
+
+        if legacy:
+            raise ValueError(
+                "Legacy solvers are not supported any longer. Please use one of"
+                " the optlang solver interfaces instead.")
+
+        if num_cpu is None:
+            num_cpu = kwargs.get("number_of_processes")
+        if num_cpu is None:
+            try:
+                num_cpu = multiprocessing.cpu_count()
+            except NotImplementedError:
+                warn("Number of cores could not be detected - assuming 1.")
+                num_cpu = 1
+
+        # Reactions with 0 flux in the current conditions have no impact on
+        # deletion. Determine those reactions now to speed up computation later.
+        if method == 'moma':
+            moma.add_moma(model)
+        if method == 'linear moma':
+            moma.add_moma(model, linear=True)
+        solution = model.optimize()
+        if solution.status == "optimal":
+            kwargs["growth"] = solution.objective_value
+            kwargs["zero_flux_reactions"] = \
+                {rxn for (rxn, flux) in solution.fluxes.items()
+                 if abs(flux) < zero_cutoff}
+        else:
+            warn("Non-optimal ({0:s}) base solution.".format(solution.status))
+
+        def unordered_combination(reactions):
+            return frozenset([i.id for i in reactions if i.id not in kwargs["zero_flux_reactions"]])
+
+        args = set([unordered_combination(comb) for comb in product(*element_lists)])
+        worker_function.model = model
+        if num_cpu > 1:
+            chunk_size = len(args) // num_cpu
+            pool = multiprocessing.Pool(num_cpu)
+            results = pool.imap_unordered(worker_function, args, chunksize=chunk_size)
+        else:
+            results = map(worker_function, args)
+        knockout_results = {frozenset(ids): 0.0
+            if growth and abs(growth) < zero_cutoff else growth for (growth, ids) in results}
+        double = infinite_defaultdict()
+        for comb in product(*element_lists):
+            growth = knockout_results[unordered_combination(comb)]
+            current = double
+            for v in comb[:-1]:
+                current = current[v.id]
+            current[comb[-1].id] = growth
+        return defaultdict_to_dict(double)
+
+
+def _element_lists(entities, *ids):
```

Comment:
In the various functions you're converting between identifiers and objects quite a lot (it seems). Is it possible to be more consistent, for example, work with identifiers most of the time unless reactions are needed for knockouts?

Reply:
the current interface for the knockout functions allows to accept both ids and objects, and this ambiguity propagated down here. But I agree there are maybe some unnecessary ones that could be cleaned up